File: matlab/src/BatOptimizer.m
classdef BatOptimizer < Solver
    %{
    Bat Algorithm Optimizer implementation.
    
    The Bat Algorithm is a metaheuristic optimization algorithm inspired by 
    the echolocation behavior of microbats. It uses frequency tuning, loudness,
    and pulse emission rate to control the search process.
    
    Parameters:
    -----------
    objective_func : function handle
        Objective function to optimize
    lb : float or array
        Lower bounds for variables
    ub : float or array
        Upper bounds for variables  
    dim : int
        Problem dimension
    maximize : bool, optional
        Optimization direction, default is True (maximize)
    varargin : cell array
        Additional algorithm parameters:
        - fmin: Minimum frequency (default: 0)
        - fmax: Maximum frequency (default: 2)
        - alpha: Loudness decay constant (default: 0.9)
        - gamma: Pulse rate increase constant (default: 0.9)
        - ro: Initial pulse emission rate (default: 0.5)
    %}
    
    properties
        fmin
        fmax
        alpha
        gamma
        ro
    end
    
    methods
        function obj = BatOptimizer(objective_func, lb, ub, dim, maximize, varargin)
            %{
            BatOptimizer constructor - Initialize the Bat Optimizer
            
            Inputs:
                objective_func : function handle
                    Objective function to optimize
                lb : float or array
                    Lower bounds of search space
                ub : float or array
                    Upper bounds of search space
                dim : int
                    Number of dimensions in the problem
                maximize : bool
                    Whether to maximize (true) or minimize (false) objective
                varargin : cell array
                    Additional solver parameters
            %}
            
            % Call parent constructor
            obj@Solver(objective_func, lb, ub, dim, maximize, varargin{:});
            
            % Set solver name
            obj.name_solver = "Bat Optimizer";
            
            % Set default BAT parameters
            obj.fmin = obj.get_kw('fmin', 0.0);          % Minimum frequency
            obj.fmax = obj.get_kw('fmax', 2.0);          % Maximum frequency
            obj.alpha = obj.get_kw('alpha', 0.9);        % Loudness decay constant
            obj.gamma = obj.get_kw('gamma', 0.9);        % Pulse rate increase constant
            obj.ro = obj.get_kw('ro', 0.5);              % Initial pulse emission rate
        end
        
        function population = init_population(obj, search_agents_no)
            %{
            init_population - Initialize population of bats with bat-specific parameters
            
            Inputs:
                search_agents_no : int
                    Number of bats to initialize
                    
            Returns:
                population : cell array
                    Cell array of initialized BatMember objects
            %}
            
            population = cell(1, search_agents_no);
            for i = 1:search_agents_no
                position = rand(1, obj.dim) .* (obj.ub - obj.lb) + obj.lb;
                fitness = obj.objective_func(position);
                % Initialize with default bat parameters
                population{i} = BatMember(...
                    position, ...
                    fitness, ...
                    0.0, ...                    % frequency
                    zeros(1, obj.dim), ...      % velocity
                    1.0, ...                    % loudness
                    obj.ro ...                  % pulse_rate
                );
            end
        end
        
        function [history_step_solver, best_solver] = solver(obj, search_agents_no, max_iter)
            %{
            solver - Main optimization method for the Bat Algorithm
            
            Parameters:
            -----------
            search_agents_no : int
                Number of bats (search agents)
            max_iter : int
                Maximum number of iterations
                
            Returns:
            --------
            Tuple containing:
                - history_step_solver: Cell array of best solutions at each iteration
                - best_solver: Best solution found overall
            %}
            
            % Initialize storage variables
            history_step_solver = {};
            
            % Initialize population
            population = obj.init_population(search_agents_no);
            
            % Initialize best solution
            [sorted_population, ~] = obj.sort_population(population);
            best_solver = sorted_population{1}.copy();
            
            % Call the begin function
            obj.begin_step_solver(max_iter);
            
            % Main optimization loop
            for iter = 1:max_iter
                % Update each bat
                for i = 1:search_agents_no
                    % Update frequency
                    population{i}.frequency = obj.fmin + (obj.fmax - obj.fmin) * rand();
                    
                    % Update velocity
                    population{i}.velocity = population{i}.velocity + ...
                        (population{i}.position - best_solver.position) * population{i}.frequency;
                    
                    % Update position
                    new_position = population{i}.position + population{i}.velocity;
                    
                    % Apply boundary constraints
                    new_position = max(min(new_position, obj.ub), obj.lb);
                    
                    % Random walk with probability (1 - pulse_rate)
                    if rand() > population{i}.pulse_rate
                        % Generate random walk step
                        epsilon = -1 + 2 * rand();
                        % Calculate mean loudness of all bats
                        loudness_values = zeros(1, search_agents_no);
                        for j = 1:search_agents_no
                            loudness_values(j) = population{j}.loudness;
                        end
                        mean_loudness = mean(loudness_values);
                        new_position = best_solver.position + epsilon * mean_loudness;
                        new_position = max(min(new_position, obj.ub), obj.lb);
                    end
                    
                    % Evaluate new fitness
                    new_fitness = obj.objective_func(new_position);
                    
                    % Create temporary BatMember for comparison
                    new_bat = BatMember(...
                        new_position, ...
                        new_fitness, ...
                        population{i}.frequency, ...
                        population{i}.velocity, ...
                        population{i}.loudness, ...
                        population{i}.pulse_rate ...
                    );
                    
                    % Update if solution improves and meets loudness criteria
                    if obj.is_better(new_bat, population{i}) && rand() < population{i}.loudness
                        % Update position and fitness
                        population{i}.position = new_position;
                        population{i}.fitness = new_fitness;
                        
                        % Update loudness and pulse rate
                        population{i}.loudness = obj.alpha * population{i}.loudness;
                        population{i}.pulse_rate = obj.ro * (1 - exp(-obj.gamma * iter));
                    end
                    
                    % Update best solution if improved
                    if obj.is_better(population{i}, best_solver)
                        best_solver = population{i}.copy();
                    end
                end
                
                % Save history
                history_step_solver{end+1} = best_solver.copy();
                
                % Call the callbacks
                obj.callbacks(iter, max_iter, best_solver);
            end
            
            % Final evaluation and storage
            obj.history_step_solver = history_step_solver;
            obj.best_solver = best_solver;
            
            % Call the end function
            obj.end_step_solver();
        end
        
        function [sorted_population, sorted_indices] = sort_population(obj, population)
            %{
            sort_population - Sort population based on fitness
            
            Inputs:
                population : cell array
                    Population to sort
                    
            Returns:
                sorted_population : cell array
                    Sorted population
                sorted_indices : array
                    Indices of sorted order
            %}
            
            % Extract fitness values from population
            fitness_values = zeros(1, length(population));
            for i = 1:length(population)
                fitness_values(i) = population{i}.fitness;
            end
            
            % Sort indices based on optimization direction
            if obj.maximize
                % Sort in descending order for maximization
                [~, sorted_indices] = sort(fitness_values, 'descend');
            else
                % Sort in ascending order for minimization
                [~, sorted_indices] = sort(fitness_values, 'ascend');
            end
            
            % Sort population based on sorted indices
            sorted_population = cell(1, length(population));
            for i = 1:length(population)
                sorted_population{i} = population{sorted_indices(i)};
            end
        end
    end
end

File: src/bat_optimizer.py
import numpy as np
from typing import Callable, Union, Tuple, List
from ._core import Solver, Member
from ._general import sort_population


class BatMember(Member):
    """
    Custom Member class for Bat Algorithm with bat-specific attributes.
    
    Attributes:
    -----------
    position : np.ndarray
        Position of the bat in search space
    fitness : float
        Fitness value of the position
    frequency : float
        Frequency used for velocity update
    velocity : np.ndarray
        Velocity vector of the bat
    loudness : float
        Loudness parameter (controls acceptance of new solutions)
    pulse_rate : float
        Pulse emission rate (controls random walk probability)
    """
    
    def __init__(self, position: np.ndarray, fitness: float, frequency: float = 0.0, 
                 velocity: np.ndarray = None, loudness: float = 1.0, pulse_rate: float = 0.5):
        super().__init__(position, fitness)
        self.frequency = frequency
        self.velocity = velocity if velocity is not None else np.zeros_like(position)
        self.loudness = loudness
        self.pulse_rate = pulse_rate
    
    def copy(self):
        return BatMember(
            self.position.copy(), 
            self.fitness, 
            self.frequency,
            self.velocity.copy(),
            self.loudness,
            self.pulse_rate
        )
    
    def __str__(self):
        return (f"Position: {self.position} - Fitness: {self.fitness} - "
                f"Frequency: {self.frequency:.3f} - Loudness: {self.loudness:.3f} - "
                f"Pulse Rate: {self.pulse_rate:.3f}")


class BatOptimizer(Solver):
    """
    Bat Algorithm Optimizer implementation.
    
    The Bat Algorithm is a metaheuristic optimization algorithm inspired by 
    the echolocation behavior of microbats. It uses frequency tuning, loudness,
    and pulse emission rate to control the search process.
    
    Parameters:
    -----------
    objective_func : Callable
        Objective function to optimize
    lb : Union[float, np.ndarray]
        Lower bounds for variables
    ub : Union[float, np.ndarray]
        Upper bounds for variables  
    dim : int
        Problem dimension
    maximize : bool, optional
        Optimization direction, default is True (maximize)
    **kwargs
        Additional algorithm parameters:
        - fmin: Minimum frequency (default: 0)
        - fmax: Maximum frequency (default: 2)
        - alpha: Loudness decay constant (default: 0.9)
        - gamma: Pulse rate increase constant (default: 0.9)
        - ro: Initial pulse emission rate (default: 0.5)
    """
    
    def __init__(self, objective_func: Callable, lb: Union[float, np.ndarray], 
                 ub: Union[float, np.ndarray], dim: int, maximize: bool = True, **kwargs):
        super().__init__(objective_func, lb, ub, dim, maximize, **kwargs)
        
        # Store additional parameters for later use
        self.kwargs = kwargs
        
        # Set solver name
        self.name_solver = "Bat Optimizer"
        
        # Set default BAT parameters
        self.fmin = kwargs.get('fmin', 0.0)          # Minimum frequency
        self.fmax = kwargs.get('fmax', 2.0)          # Maximum frequency
        self.alpha = kwargs.get('alpha', 0.9)        # Loudness decay constant
        self.gamma = kwargs.get('gamma', 0.9)        # Pulse rate increase constant
        self.ro = kwargs.get('ro', 0.5)              # Initial pulse emission rate
    
    def _init_population(self, search_agents_no) -> List:
        """
        Initialize population of bats with bat-specific parameters.
        
        Parameters:
        -----------
        search_agents_no : int
            Number of bats to initialize
            
        Returns:
        --------
        List[BatMember]
            List of initialized bat members
        """
        population = []
        for _ in range(search_agents_no):
            position = np.random.uniform(self.lb, self.ub, self.dim)
            fitness = self.objective_func(position)
            # Initialize with default bat parameters
            population.append(BatMember(
                position, 
                fitness, 
                frequency=0.0,
                velocity=np.zeros(self.dim),
                loudness=1.0,
                pulse_rate=self.ro
            ))
        return population

    def solver(self, search_agents_no: int, max_iter: int) -> Tuple[List, BatMember]:
        """
        Main optimization method for the Bat Algorithm.
        
        Parameters:
        -----------
        search_agents_no : int
            Number of bats (search agents)
        max_iter : int
            Maximum number of iterations
            
        Returns:
        --------
        Tuple[List, BatMember]
            History of best solutions and the best solution found
        """
        # Initialize population
        population = self._init_population(search_agents_no)
        
        # Find initial best solution
        sorted_population, _ = self._sort_population(population)
        best_solver = sorted_population[0].copy()
        
        # Initialize history
        history_step_solver = []
        
        # Start solver (show progress bar)
        self._begin_step_solver(max_iter)
        
        # Main optimization loop
        for iter in range(max_iter):
            # Update each bat
            for i in range(search_agents_no):
                # Update frequency
                population[i].frequency = self.fmin + (self.fmax - self.fmin) * np.random.random()
                
                # Update velocity
                population[i].velocity = population[i].velocity + (population[i].position - best_solver.position) * population[i].frequency
                
                # Update position
                new_position = population[i].position + population[i].velocity
                
                # Apply boundary constraints
                new_position = np.clip(new_position, self.lb, self.ub)
                
                # Random walk with probability (1 - pulse_rate)
                if np.random.random() > population[i].pulse_rate:
                    # Generate random walk step
                    epsilon = -1 + 2 * np.random.random()
                    # Calculate mean loudness of all bats
                    mean_loudness = np.mean([bat.loudness for bat in population])
                    new_position = best_solver.position + epsilon * mean_loudness
                    new_position = np.clip(new_position, self.lb, self.ub)
                
                # Evaluate new fitness
                new_fitness = self.objective_func(new_position)
                
                # Create temporary BatMember for comparison
                new_bat = BatMember(new_position, new_fitness, population[i].frequency, 
                                   population[i].velocity.copy(), population[i].loudness, population[i].pulse_rate)
                
                # Update if solution improves and meets loudness criteria
                if (self._is_better(new_bat, population[i]) and 
                    np.random.random() < population[i].loudness):
                    
                    # Update position and fitness
                    population[i].position = new_position
                    population[i].fitness = new_fitness
                    
                    # Update loudness and pulse rate
                    population[i].loudness = self.alpha * population[i].loudness
                    population[i].pulse_rate = self.ro * (1 - np.exp(-self.gamma * iter))
                
                # Update best solution if improved
                if self._is_better(population[i], best_solver):
                    best_solver = population[i].copy()
            
            # Save history
            history_step_solver.append(best_solver.copy())
            
            # Call callback for progress tracking
            self._callbacks(iter, max_iter, best_solver)
        
        # Store final results
        self.history_step_solver = history_step_solver
        self.best_solver = best_solver
        
        # End solver
        self._end_step_solver()
        
        return history_step_solver, best_solver

    def _sort_population(self, population):
        return sort_population(population, self.maximize)